### PR TITLE
David knigge/diffusion experiment

### PR DESCRIPTION
**Pull request by david-knigge**
Implemented a first-pass of diffusion generative modelling using HuggingFace's diffusers package. 

- Generalized resnet implementation for pixel-level output with added AdaLN residual block for conditioning on timestep/label.
- By default we do DDPM with a linear beta schedule. 
- We track the model EMA for sampling.
- Also implemented support for classifier-free guidance. 

Verified using small-scale experiment on MNIST, see attached results.

image

Todo:
- Implement FID
- Implement Simple Diffusion 2 tricks; schedule and sigmoid loss-weighting.

### CODE REVIEW

File: experiments/datamodules/imagenet.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+            (DEFAULT_IMAGENET_MEAN, DEFAULT_IMAGENET_STD),
+        )
+        self.normalization_mean = tuple(float(m) for m in mean)
+        self.normalization_std = tuple(float(s) for s in std)
+
+        self.train_dataset: Optional[_ImageNetDataset] = None
+        self.val_dataset: Optional[_ImageNetDataset] = None
+
+    def _build_transform(self, *, train: bool) -> transforms.Compose:
+        ops: list[transforms.Compose | transforms.RandomCrop | transforms.CenterCrop | transforms.Resize | transforms.RandomHorizontalFlip | transforms.ToTensor] = [
```
Comment (Dafidofff) [inline experiments/datamodules/imagenet.py:140]: Are these the transforms used for simple-diffusion2? If so, maybe add an comment here to make it explicit

**Review by Dafidofff — APPROVED** [review]

File: examples/imagenet_diffusion/ccnn_12_768_hyena_rope_qknorm.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+            mlp_norm_cfg="${net.norm_cfg}",
+            dropout_cfg=LazyConfig(torch.nn.Dropout)(p=DROPOUT_RATE),
+            condition_norm_cfg="${net.norm_cfg}",
+            hidden_dim="${net.hidden_dim}",
+        ),
+        dropout_in_cfg=LazyConfig(torch.nn.Dropout)(p=DROPOUT_IN_RATE),
+        condition_in_proj_cfg=LazyConfig(torch.nn.Linear)(in_features=PLACEHOLDER, out_features=PLACEHOLDER),
+    )
+
+    config.lightning_wrapper_class = LazyConfig(DiffusionWrapper)()
```
Comment (dwromero) [inline examples/imagenet_diffusion/ccnn_12_768_hyena_rope_qknorm.py:175]: @david-knigge should we put the Diffusion config in the wrapper definition?

File: experiments/datamodules/__init__.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+from .imagenet import ImageNetDataModule  # noqa: F401
+from .mnist import MNISTDataModule  # noqa: F401
```
Comment (dwromero) [inline experiments/datamodules/__init__.py:2]: Let's please do not do this. Let's access datasets by entire file.

__init__ should be empty, and we should import with complete paths, not relative ones.

File: experiments/trainer.py [hunk 4]
```diff
@@ -92,23 +100,27 @@ def construct_trainer(
         *user_callbacks,
     ]
 
-    # create trainer
-    trainer = pl.Trainer(
-        max_steps=cfg.train.iterations,
-        logger=wandb_logger,
-        gradient_clip_val=cfg.train.grad_clip,
-        accumulate_grad_batches=cfg.train.accumulate_grad_steps,
+    trainer_kwargs: dict[str, object] = {
```
Comment (dwromero) [inline experiments/trainer.py]: Any reason for these changes?

File: experiments/utils/cli.py [hunk 5]
```diff
@@ -109,10 +112,15 @@ def load_config_from_file(config_path: str) -> ExperimentConfig:
     if module_path.endswith(".py"):
         module_path = module_path[:-3]  # Remove .py extension
 
-    # Import the module
-    spec = importlib.util.spec_from_file_location(module_path, config_path)
-    module = importlib.util.module_from_spec(spec)
-    spec.loader.exec_module(module)
+    # Import the module. Prefer a regular import so package-relative references
```
Comment (dwromero) [inline experiments/utils/cli.py]: Let's please not use relative imports. That makes things error-rpone

File: experiments/utils/cli.py [hunk 6]
```diff
@@ -109,10 +112,15 @@ def load_config_from_file(config_path: str) -> ExperimentConfig:
...
+    # provided path if the module isn't importable (e.g., outside PYTHONPATH).
+    try:
+        module = importlib.import_module(module_path)
+    except ModuleNotFoundError:
+        spec = importlib.util.spec_from_file_location(module_path, config_path)
+        module = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(module)
 
     # Get the get_config function
     if not hasattr(module, "get_config"):
```
Comment (dwromero) [inline experiments/utils/cli.py:120]: This is a hack? We should remove this.

File: examples/imagenet_diffusion/ccnn_12_768_hyena_rope_qknorm.py (hunk 2 above)
Comment (david-knigge) [reply examples/imagenet_diffusion/ccnn_12_768_hyena_rope_qknorm.py:175]: We definitely could, but we're starting to get a pretty big set of diffusion config args, so for maintainability it might be better to have the single location in default_cfg.py to touch when we're modifying the config, or i might start missing args for the different experiments we're doing :))

File: experiments/utils/cli.py (hunk 6 above)
Comment (david-knigge) [reply experiments/utils/cli.py:120]: Done! Codex got antsy

File: experiments/trainer.py (hunk 4 above)
Comment (david-knigge) [reply experiments/trainer.py]: Nope, just cosmetics, rolled them back

File: experiments/datamodules/imagenet.py (hunk 1 above)
Comment (david-knigge) [reply experiments/datamodules/imagenet.py:140]: Yes took these from the appendix.
